libraries/rust/data/src/client.rs

```diff
@@ -0,0 +1,218 @@
+use chrono::{DateTime, Utc};
+use reqwest::Url;
+use reqwest::header::{HeaderMap, HeaderValue, ACCEPT};
+use reqwest::Client as HTTPClient;
+use aws_sdk_s3::client::Client as S3Client;
+use aws_credential_types::Credentials;
+use aws_types::region::Region;
+use aws_sdk_s3::Config;
+use serde::{Serialize, Deserialize};
+use serde_json;
+use aws_sdk_s3::primitives::ByteStream;
+use flate2::write::GzEncoder;
+use flate2::read::GzDecoder;
+use flate2::Compression;
+use std::io::{Write, Read};
+
+#[derive(Deserialize)]
+struct BarsResponse {
+    bars: Vec<Bar>,
+    #[serde(rename = "next_page_token")]
+    next_token: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Bar {
+    pub ticker: Option<String>,
+    #[serde(rename = "t")]
+    pub timestamp: DateTime<Utc>,
+    #[serde(rename = "o")]
+    pub open: f32,
+    #[serde(rename = "h")]
+    pub high: f32,
+    #[serde(rename = "l")]
+    pub low: f32,
+    #[serde(rename = "c")]
+    pub close: f32,
+    #[serde(rename = "v")]
+    pub volume: u32,
+    #[serde(rename = "n")]
+    pub number_of_trades: u32,
+    #[serde(rename = "vw")]
+    pub volume_weighted_average_price: f32,
+}
+
+pub struct Client {
+    alpaca_base_url: String,
+    alpaca_api_key_id: String,
+    alpaca_api_secret_key: String,
+    http_client: HTTPClient,
+    s3_client: S3Client,
+    s3_data_bucket_name: String,
+}
+
+const EQUITY_BARS_PATH: &str = "equity/bars";
+
+impl Client {
+    pub fn new(
+        alpaca_api_key_id: String,
+        alpaca_api_secret_key: String,
+        aws_access_key_id: String,
+        aws_secret_access_key: String,
+        s3_data_bucket_name: String,
+    ) -> Self {
+        let creds = Credentials::from_keys(aws_access_key_id, aws_secret_access_key, None);
+        
+        let conf = Config::builder()
+            .credentials_provider(creds)
+            .region(Region::new("us-east-1"))
+            .build();
+        
+        let s3_client = S3Client::from_conf(conf);
+
+        Client {
+            alpaca_base_url: "https://data.alpaca.markets/v2/".to_string(),
+            alpaca_api_key_id,
+            alpaca_api_secret_key,
+            http_client: HTTPClient::new(),
+            s3_client,
+            s3_data_bucket_name,
+        }
+    }
+
+    pub async fn fetch_equities_bars_by_date_range(
+        &self,
+        tickers: Vec<String>,
+        start: DateTime<Utc>,
+        end: DateTime<Utc>,
+    ) -> Result<Vec<Bar>, Box<dyn std::error::Error>> {
+        let mut equities_bars: Vec<Bar> = Vec::new();
+
+        let mut headers = HeaderMap::new();
+        headers.insert(ACCEPT, HeaderValue::from_str("application/json")?);
+        headers.insert("APCA-API-KEY-ID", HeaderValue::from_str(&self.alpaca_api_key_id)?);
+        headers.insert("APCA-API-SECRET-KEY", HeaderValue::from_str(&self.alpaca_api_secret_key)?);
+
+        for ticker in tickers {
```

Comment:
I think this is probably good enough, just making a note that we should monitor the performance of this and use tokio to spawn threads if needed

Reply:
Agreed this is currently a brute force implementation and highly inefficient.